srml/balances/src/lib.rs

```diff
@@ -0,0 +1,1083 @@
+	type Origin = T::Origin;
+	type Call = T::Call;
+	type Index = T::Index;
+	type BlockNumber = T::BlockNumber;
+	type Hash = T::Hash;
+	type Hashing = T::Hashing;
+	type AccountId = T::AccountId;
+	type Lookup = T::Lookup;
+	type Header = T::Header;
+	type Event = ();
+	type BlockHashCount = T::BlockHashCount;
+	type MaximumBlockWeight = T::MaximumBlockWeight;
+	type MaximumBlockLength = T::MaximumBlockLength;
+	type AvailableBlockRatio = T::AvailableBlockRatio;
+	type Version = T::Version;
+}
+impl<T: Subtrait<I>, I: Instance> timestamp::Trait for ElevatedTrait<T, I> {
+	type Moment = T::Moment;
+	type OnTimestampSet = ();
+	type MinimumPeriod = T::MinimumPeriod;
+}
+impl<T: Subtrait<I>, I: Instance> Trait<I> for ElevatedTrait<T, I> {
+	type Balance = T::Balance;
+	type OnFreeBalanceZero = T::OnFreeBalanceZero;
+	type OnNewAccount = T::OnNewAccount;
+	type TransferPayment = ();
+	type DustRemoval = ();
+	type Event = ();
+	type ExistentialDeposit = T::ExistentialDeposit;
+	type TransferFee = T::TransferFee;
+	type CreationFee = T::CreationFee;
+}
+
+impl<T: Trait<I>, I: Instance> Currency<T::AccountId> for Module<T, I>
+where
+	T::Balance: MaybeSerializeDeserialize + Debug,
+{
+	type Balance = T::Balance;
+	type PositiveImbalance = PositiveImbalance<T, I>;
+	type NegativeImbalance = NegativeImbalance<T, I>;
+
+	fn total_balance(who: &T::AccountId) -> Self::Balance {
+		Self::free_balance(who) + Self::reserved_balance(who)
+	}
+
+	fn can_slash(who: &T::AccountId, value: Self::Balance) -> bool {
+		Self::free_balance(who) >= value
+	}
+
+	fn total_issuance() -> Self::Balance {
+		<TotalIssuance<T, I>>::get()
+	}
+
+	fn minimum_balance() -> Self::Balance {
+		T::ExistentialDeposit::get()
+	}
+
+	fn burn(mut amount: Self::Balance) -> Self::PositiveImbalance {
+		<TotalIssuance<T, I>>::mutate(|issued| {
+			*issued = issued.checked_sub(&amount).unwrap_or_else(|| {
+				amount = *issued;
+				Zero::zero()
+			});
+		});
+		PositiveImbalance::new(amount)
+	}
+
+	fn issue(mut amount: Self::Balance) -> Self::NegativeImbalance {
+		<TotalIssuance<T, I>>::mutate(|issued| {
+			*issued = issued.checked_add(&amount).unwrap_or_else(|| {
+				amount = Self::Balance::max_value() - *issued;
+				Self::Balance::max_value()
+			})
+		});
+		NegativeImbalance::new(amount)
+	}
+
+	fn free_balance(who: &T::AccountId) -> Self::Balance {
+		<FreeBalance<T, I>>::get(who)
+	}
+
+	// # <weight>
+	// Despite iterating over a list of locks, they are limited by the number of
+	// lock IDs, which means the number of runtime modules that intend to use and create locks.
+	// # </weight>
+	fn ensure_can_withdraw(
+		who: &T::AccountId,
+		amount: T::Balance,
+		reasons: WithdrawReasons,
+		new_balance: T::Balance,
+	) -> Result {
+		if reasons.intersects(WithdrawReason::Reserve | WithdrawReason::Transfer)
+			&& Self::vesting_balance(who) > new_balance
+		{
+			return Err("vesting balance too high to send value");
+		}
+		let (staking_lock, unbonding_locks) = Self::locks(who);
```

Comment:
hmm... why not put staking lock and unbonding_lock in one lock, it is not a good idea that the balances module dependent on staking.

Reply:
All **Unbonding Locks** share one **Staking Lock**. If you define a lock which have a `staking_amount` filed and it also depends on staking logic. In substrate the `BalanceLock` was defined in balances module. There're two kinds of Currency on our chain, so I move the `BalanceLock to support module.

Reply:
我觉得Lockable Currency的lock机制不只是Staking模块会用到，其他模块也会用到。因此Balance模块中的Lock功能应该通用，特别的业务逻辑应该放到对应的业务模块中去：这里的staking_lock，unbounding_lock及其判断逻辑应该放在staking模块或support模块。

所以具体的结构还好，主要是这些特殊的lock具体实现机制不要放在Balance模块中

Reply:
There're some problems with your `check_valid`.

1. `new_balance >= staking_amount + (all)unbonding_amount`
2. `new_balance >= staking_amount` only need to check once

How about the `CompositeLock`?

```rust
CompositLock {
	staking_amount: Balance,
	locks: Vec<BalanceLock>,
}
```

Reply:
Yes, this is better. (I forgot that there are multiple unbounding locks in staking).

And I suggest move all the checks to Lock trait(enum), CompositLock will require to implement 
```
check_valid(new_blance, time) && reasons.check ...
```
Then simply call lock.check_valid(new_blance, time) regardless what kind of it is in ```ensure withdraw```

Reply:
Yep, I wanted to move the whole `ensure_can_withdraw` logic into `LockableCurrency` at the beginning. It would be more concise.

And there's a [weird problem](https://github.com/darwinia-network/darwinia/pull/94#issuecomment-555331777) when I try to merge from develop branch. I can't compile the staking module but I can compile the darwinia.

Request review:
- [LockableCurrency](https://github.com/AurevoirXavier/darwinia/blob/9554a520375f7e0fd1546780e3322739b86a6734/srml/support/src/traits.rs#L25)
- [staking](https://github.com/AurevoirXavier/darwinia/blob/chainrelay/srml/staking/src/lib.rs)
- [balances](https://github.com/AurevoirXavier/darwinia/blob/chainrelay/srml/balances/src/lib.rs)
- [kton](https://github.com/AurevoirXavier/darwinia/blob/chainrelay/srml/kton/src/lib.rs)